src/System.Web.OData/OData/Routing/DefaultODataPathHandler.cs

```diff
@@ -117,6 +117,21 @@ private ODataPath Parse(string serviceRoot, string odataPath, IServiceProvider r
                         ? serviceRoot
                         : serviceRoot + "/");
 
+                // Special handling for the cases when odataPath is invalid.
+                if (!Uri.IsWellFormedUriString(odataPath, UriKind.RelativeOrAbsolute))
+                {
+                    // Due to a bug in the System.Uri some relative paths are rejected if they contain
+                    // a ':' symbol on a position greater than 1024. This careful check should mitigate 
+                    // this problem by encoding these characters before the path is combined with 
+                    // service roor Uri.
+                    // https://github.com/dotnet/corefx/issues/29011
+                    var odataPathColonEncoded = odataPath.Replace(":", "%3A");
```

Comment:
Could you add check for length? Replace will generate new string => more memory and GC pressure.